src/primitives/gf32.rs

```diff
@@ -0,0 +1,481 @@
+// Written by the Andrew Poelstra and the rust-bitcoin developers.
+// (Replacing very similar code originally written by Clark Moody.)
+// SPDX-License-Identifier: MIT
+
+//! GF32 - Galois Field over 32 elements.
+//!
+//! Implements GF32 arithmetic, defined and encoded as in BIP-0173 "bech32".
+//!
+//! > A finite field is a finite set which is a field; this means that multiplication, addition,
+//! > subtraction and division (excluding division by zero) are defined and satisfy the rules of
+//! > arithmetic known as the field axioms.
+//!
+//! ref: <https://en.wikipedia.org/wiki/Finite_field>
+
+use core::convert::{Infallible, TryFrom};
+use core::{fmt, num, ops, str};
+
+#[cfg(all(test, mutate))]
+use mutagen::mutate;
+
+/// Logarithm table of each bech32 element, as a power of alpha = Z.
+///
+/// Includes Q as 0 but this is false; you need to exclude Q because it has no discrete log. If we
+/// could have a 1-indexed array that would panic on a 0 index that would be better.
+#[rustfmt::skip]
+const LOG: [isize; 32] = [
+     0,  0,  1, 14,  2, 28, 15, 22,
+     3,  5, 29, 26, 16,  7, 23, 11,
+     4, 25,  6, 10, 30, 13, 27, 21,
+    17, 18,  8, 19, 24,  9, 12, 20,
+];
+
+/// Mapping of powers of 2 to the numeric value of the element.
+#[rustfmt::skip]
+const LOG_INV: [u8; 31] = [
+     1,  2,  4,  8, 16,  9, 18, 13,
+    26, 29, 19, 15, 30, 21,  3,  6,
+    12, 24, 25, 27, 31, 23,  7, 14,
+    28, 17, 11, 22,  5, 10, 20,
+];
+
+/// Mapping from numeric value to bech32 character.
+#[rustfmt::skip]
+const CHARS_LOWER: [char; 32] = [
+    'q', 'p', 'z', 'r', 'y', '9', 'x', '8', //  +0
+    'g', 'f', '2', 't', 'v', 'd', 'w', '0', //  +8
+    's', '3', 'j', 'n', '5', '4', 'k', 'h', // +16
+    'c', 'e', '6', 'm', 'u', 'a', '7', 'l', // +24
+];
+
+/// Mapping from bech32 character (either case) to numeric value.
+#[rustfmt::skip]
+const CHARS_INV: [i8; 128] = [
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    15, -1, 10, 17, 21, 20, 26, 30,  7,  5, -1, -1, -1, -1, -1, -1,
+    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1,
+    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1,
+];
+
+/// An element of GF32.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Fe32(pub(crate) u8); // pub(crate) because of `hrpstring` module.
```

Comment:
In 310012dd3e5f5725b5bde9021606ed1d5542620d:

I'd remove this comment. Every module in `primitives` will probably use the internal `u8`.